colsample_bytree seemingly not random

My understanding of `colsample_bytree` is that it randomly samples from the features (columns) for each tree, so it would be a way to limit dimensionality prior to constructing trees. However when I use values < 1, it seems to always take from the first columns (based on whatever percentage) and not randomly. Extending on the sample R code for xgb.train:

```
data(agaricus.train, package='xgboost')
data(agaricus.test, package='xgboost')

dtrain <- xgb.DMatrix(agaricus.train$data, label = agaricus.train$label)
dtest <- xgb.DMatrix(agaricus.test$data, label = agaricus.test$label)
watchlist <- list(train = dtrain, eval = dtest)

## A simple xgb.train example:
param <- list(max_depth = 2, eta = 1, silent = 1, nthread = 2, colsample_bytree = 1,
              objective = "binary:logistic", eval_metric = "auc")
bst <- xgb.train(param, dtrain, nrounds = 100, watchlist, verbose = 0)
xgb.importance(model = bst)

param <- list(max_depth = 2, eta = 1, silent = 1, nthread = 2, colsample_bytree = 0.01,
              objective = "binary:logistic", eval_metric = "auc")
bst <- xgb.train(param, dtrain, nrounds = 100, watchlist, verbose = 0)
xgb.importance(model = bst)
```
0.01 is equivalent to saying 1 feature and the last importance matrix always shows `cap-shape=bell`, the first column in the data. I wouldn't expect this if the column sampling were random. I see this with my own data too, where X for colsample_bytree always results in columns in the importance matrix being from the first X * 100% of columns in my data matrix. So if someone uses 0.8, they would essentially be throwing away the last 20% of their columns and I don't think that's what users would want or expect they're doing.

@Laurae2 I get the same thing as @ben519 when running your code. Including:

```
> mean(unlist(lapply(1:1000, function(x) {
+   set.seed(x)
+   bst <- xgb.train(param, dtrain, nrounds = 100, watchlist, verbose = 0)
+   feat <- xgb.importance(model = bst)
+   mean(as.numeric(substr(feat$Feature, start = 9, stop = 11)))
+ })))
[1] 1
> table(unlist(lapply(1:10, function(x) {
+   set.seed(x)
+   bst <- xgb.train(param, dtrain, nrounds = 10000, watchlist, verbose = 0)
+   feat <- xgb.importance(model = bst)
+   as.numeric(substr(feat$Feature, start = 9, stop = 11))
+ })))

 1 
10 
```
Regular R randomness is fine:
```
> table(unlist(lapply(1:10, function(x) {
+   set.seed(x)
+   unique(sample(1:100, size = 10000, replace = TRUE))
+ })))

  1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17  18  19  20  21  22  23  24  25  26  27  28  29  30 
 10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10 
 31  32  33  34  35  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53  54  55  56  57  58  59  60 
 10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10 
 61  62  63  64  65  66  67  68  69  70  71  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89  90 
 10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10  10 
 91  92  93  94  95  96  97  98  99 100 
 10  10  10  10  10  10  10  10  10  10 
```

```
> Sys.info()
                                                                                          sysname 
                                                                                         "Darwin" 
                                                                                          release 
                                                                                         "17.7.0" 
                                                                                          version 
"Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64"
```
```
> sessionInfo()
R version 3.5.1 (2018-07-02)
Platform: x86_64-apple-darwin17.7.0 (64-bit)
Running under: macOS High Sierra 10.13.6

Matrix products: default
BLAS: /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /usr/local/Cellar/openblas/0.3.2/lib/libopenblas_haswellp-r0.3.2.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

loaded via a namespace (and not attached):
 [1] compiler_3.5.1    magrittr_1.5      Matrix_1.2-14     tools_3.5.1       xgboost_0.71.2    stringi_1.2.4    
 [7] grid_3.5.1        data.table_1.11.4 packrat_0.4.9-3   lattice_0.20-35
```
```
Using built-in specs.
COLLECT_GCC=gcc-8
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/8.2.0/libexec/gcc/x86_64-apple-darwin17.7.0/8.2.0/lto-wrapper
Target: x86_64-apple-darwin17.7.0
Configured with: ../configure --build=x86_64-apple-darwin17.7.0 --prefix=/usr/local/Cellar/gcc/8.2.0 --libdir=/usr/local/Cellar/gcc/8.2.0/lib/gcc/8 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-8 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --enable-checking=release --with-pkgversion='Homebrew GCC 8.2.0' --with-bugurl=https://github.com/Homebrew/homebrew-core/issues --disable-nls
Thread model: posix
gcc version 8.2.0 (Homebrew GCC 8.2.0)
```
```
❯ cat ~/.R/Makevars 

CC=/usr/local/opt/llvm/bin/clang
CXX=/usr/local/opt/llvm/bin/clang++
CFLAGS=-g -O3 -Wall -pedantic -std=gnu99 -mtune=native -pipe
CXXFLAGS=-g -O3 -Wall -pedantic -std=c++11 -mtune=native -pipe
LDFLAGS=-L/usr/local/opt/gettext/lib -L/usr/local/opt/llvm/lib -Wl,-rpath,/usr/local/opt/llvm/lib
CPPFLAGS=-I/usr/local/opt/gettext/include -I/usr/local/opt/llvm/include

❯ /usr/local/opt/llvm/bin/clang++ -v
clang version 6.0.1 (tags/RELEASE_601/final)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
InstalledDir: /usr/local/opt/llvm/bin
```
Attempting to install from source results in error:
```
> xgbdl::xgb.dl(compiler = "gcc")
Cloning into 'xgboost'...
Submodule 'cub' (https://github.com/NVlabs/cub) registered for path 'cub'
Submodule 'dmlc-core' (https://github.com/dmlc/dmlc-core) registered for path 'dmlc-core'
Submodule 'rabit' (https://github.com/dmlc/rabit) registered for path 'rabit'
Cloning into '/private/var/folders/h3/59qfw93n6k92mh9lqzvpb1480000gn/T/Rtmp9a3EWy/xgboost/cub'...
Cloning into '/private/var/folders/h3/59qfw93n6k92mh9lqzvpb1480000gn/T/Rtmp9a3EWy/xgboost/dmlc-core'...
Cloning into '/private/var/folders/h3/59qfw93n6k92mh9lqzvpb1480000gn/T/Rtmp9a3EWy/xgboost/rabit'...
Submodule path 'cub': checked out 'b20808b1b04ec3d6a625e51fbc1eb76f337754ad'
Submodule path 'dmlc-core': checked out '459ab734d15acd68fd437abf845c7c1730b5a38f'
Submodule path 'rabit': checked out '87143deb4c0a34302f727ba35497e3380b2cced8'
Already on 'master'
Your branch is up to date with 'origin/master'.
-- The C compiler identification is AppleClang 9.1.0.9020039
-- The CXX compiler identification is AppleClang 9.1.0.9020039
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES) 
-- Could NOT find OpenMP_CXX (missing: OpenMP_CXX_FLAGS OpenMP_CXX_LIB_NAMES) 
-- Could NOT find OpenMP (missing: OpenMP_C_FOUND OpenMP_CXX_FOUND) 
-- Setting build type to 'Release' as none was specified.
-- Performing Test SUPPORT_CXX11
-- Performing Test SUPPORT_CXX11 - Success
-- Performing Test SUPPORT_CXX0X
-- Performing Test SUPPORT_CXX0X - Success
-- Performing Test SUPPORT_MSSE2
-- Performing Test SUPPORT_MSSE2 - Success
CMake Error at /usr/local/Cellar/cmake/3.12.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find OpenMP_C (missing: OpenMP_C_FLAGS OpenMP_C_LIB_NAMES)
Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.12.1/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/local/Cellar/cmake/3.12.1/share/cmake/Modules/FindOpenMP.cmake:466 (find_package_handle_standard_args)
  dmlc-core/CMakeLists.txt:117 (find_package)


-- Configuring incomplete, errors occurred!
See also "/var/folders/h3/59qfw93n6k92mh9lqzvpb1480000gn/T/Rtmp9a3EWy/xgboost/build/CMakeFiles/CMakeOutput.log".
make: *** No rule to make target `install'.  Stop.
```

--- a/R-package/tests/testthat/test_basic.R
+++ b/R-package/tests/testthat/test_basic.R
@@ -237,3 +237,28 @@ test_that("max_delta_step works", {
   expect_true(all(bst1$evaluation_log$train_logloss < bst2$evaluation_log$train_logloss))
   expect_lt(mean(bst1$evaluation_log$train_logloss)/mean(bst2$evaluation_log$train_logloss), 0.8)
 })
+
+test_that("colsample_bytree works", {
+  # Randomly generate data matrix by sampling from uniform distribution [-1, 1]
+  set.seed(1)
+  train_x <- matrix(runif(1000, min = -1, max = 1), ncol = 100)
+  train_y <- as.numeric(rowSums(train_x) > 0)
+  test_x <- matrix(runif(1000, min = -1, max = 1), ncol = 100)
+  test_y <- as.numeric(rowSums(test_x) > 0)
+  colnames(train_x) <- paste0("Feature_", sprintf("%03d", 1:100))
+  colnames(test_x) <- paste0("Feature_", sprintf("%03d", 1:100))
+   dtrain <- xgb.DMatrix(train_x, label = train_y)
+  dtest <- xgb.DMatrix(test_x, label = test_y)
+  watchlist <- list(train = dtrain, eval = dtest)
+   # Use colsample_bytree = 0.01, so that roughly one out of 100 features is
+  # chosen for each tree
+  param <- list(max_depth = 2, eta = 0, silent = 1, nthread = 2,
+                colsample_bytree = 0.01, objective = "binary:logistic",
+                eval_metric = "auc")
+   set.seed(2)
+  bst <- xgb.train(param, dtrain, nrounds = 100, watchlist, verbose = 0)
+  xgb.importance(model = bst)
+  # If colsample_bytree works properly, a variety of features should be used
+  # in the 100 trees
+  expect_gte(nrow(xgb.importance(model = bst)), 30)
+})

--- a/src/common/random.h
+++ b/src/common/random.h
@@ -33,14 +33,14 @@ using RandomEngine = std::mt19937;
 class CustomGlobalRandomEngine {
  public:
   /*! \brief The result type */
-  typedef size_t result_type;
+  using result_type = uint32_t;
   /*! \brief The minimum of random numbers generated */
   inline static constexpr result_type min() {
     return 0;
   }
   /*! \brief The maximum random numbers generated */
   inline static constexpr result_type max() {
-    return std::numeric_limits<size_t>::max();
+    return std::numeric_limits<result_type>::max();
   }
   /*!
    * \brief seed function, to be implemented

